### PR TITLE
ACCUMULO-3384 support custom system props

### PR DESCRIPTION
**Pull request by ctubbsii**
Add a property prefix for custom system-wide properties

### CODE REVIEW

**Comment by joshelser** [comment]
This kind of feels like scope creep to me (why do we want to store arbitrary system metadata for users?) but whatever.

**Comment by ctubbsii** [comment]
@joshelser It's poorly worded. It's not arbitrary system metadata. It's "accumulo system"-wide configuration for pluggable components. Currently, we have a prefix for arbitrary "per-table" configuration, but that's no good for things like volume choosers, pluggable security configuration options, etc.

Having this allows pluggable components to be configurable, without changing code, or requiring pluggable components use their own configuration system.

**Comment by keith-turner** [comment]
> Having this allows pluggable components to be configurable, without changing code, or requiring pluggable components use their own configuration system.

@ctubbsii I think it would be useful to explain this in the props docs

**Comment by ctubbsii** [comment]
@keith-turner Can do. Got a few more commits I want to create PRs for, which depend on this, then I can throw a doc-update on top of that.

**Comment by keith-turner** [comment]
Would also be nice to update the javadoc for the plugins that could benefit from this prop to mention the prop.

**Comment by ctubbsii** [comment]
@phrocker Mainly, because that was not the solution discussed back in Dec2015-Jan2016 on https://issues.apache.org/jira/browse/ACCUMULO-4085 and would not follow precedent of the per-table custom configs for iterators, per-table volume choosers, etc. It's also significantly more complex, and could create chicken-and-egg problems if config is needed system-wide for scan-time components or anything that might be pluggable which is run before the tables are online.

**Comment by phrocker** [comment]
@ctubbsii I see. It seems that ACCUMULO-4085 is a specific issue that can be addressed with HDFS features. My comment was directed to the openness of this. You're suggesting the need for this for Accumulo internal features? Sorry for rehashing tickets that have been around a while, but I expect you know that is an expectation.

**Comment by ctubbsii** [comment]
@phrocker I don't understand your comment about HDFS features. This issue is completely unrelated to HDFS features, as is ACCUMULO-4085.

As I've mentioned, this is specifically to provide a cluster-wide configuration store for user-created implementations of our pluggable interfaces. VolumeChooser is one such thing, Balancer is another, iterators could use it for non-table-specific configs, the three pluggable security modules could also use it, etc.

**Comment by joshelser** [comment]
> if config is needed system-wide for scan-time components or anything that might be pluggable which is run before the tables are online.

I can't really argue against the validity of providing some prefix for yet-to-be-defined server-side extensions (by definitions, those don't yet exist so I can't say if they're a good idea). Being able to configure such deployed code is a good idea.

However, the original list of "examples" from the ticket ("list contact information for a system owner or emergency contact, for display on a monitoring site; add notes about last time a system was manually rebooted or checked for healthy status.") is what bothered me. It seems like that really isn't the intent of this change -- I'd suggest (like Keith already did) to make this clear via the Property doc (and user manual).

**Comment by ctubbsii** [comment]
I've updated the ticket description to clarify the stronger use case for this feature, and will ensure that this is documented in the property prefix description, which I can link to from the javadoc in the various pluggable interfaces.

**Comment by joshelser** [comment]
Thanks for that.

**Comment by joshelser** [comment]
@ctubbsii no user-manual update?

**Comment by ctubbsii** [comment]
@joshelser The property's own documentation is included in the user-manual appendix for properties. I didn't feel it needed a whole separate paragraph beyond that.

**Comment by joshelser** [comment]
> I didn't feel it needed a whole separate paragraph beyond that.

Wasn't suggesting its own paragraph, but I assume we have mention of instructions/descriptions around the aforementioned pluggable components (balancer, volume chooser) in which a link to the property would be beneficial (with a simple mention along the lines of "Implementations can use this property to configure blah blah"). Maybe I'm expecting too much and we don't actually have any documentation on these subjects?

**Comment by ctubbsii** [comment]
I cannot think of anything to add the user-manual which would add any value to what's already there. I think the user-manual is already too wordy, and repeating the same information in multiple places to suggest *possible* uses of a generic feature I think would be excessive, even if we did have a section in the user manual covering these non-public modular API components in detail.

**Comment by ctubbsii** [comment]
If you can think of something which will add value, by all means, I will support you... I just don't think anything else that I could say about it would be a valuable addition to the user manual.

**Comment by joshelser** [comment]
> even if we did have a section in the user manual covering these non-public modular API components in detail.

So we're adding support/features for things for the benefit of user customization that we don't actually intend to tell users how to use? We've reached another one of those impasses where I obviously don't understand your reasoning and don't care to continue re-hashing it.

**Comment by ctubbsii** [comment]
@joshelser That's not what I said at all. I didn't say I don't intend to tell users how to use it. I said that I think I sufficiently *did* explain how to use it.

I don't know what more you want from me. I added as much documentation as I felt was useful to add, and no more. I appreciate the fact that you feel it is still insufficient... but I did the best job I could with documenting it. If you want to add to it, please do so... I'm happy to support that... my feelings won't be hurt at all. But I've done the best I can with it at the moment and need to move on.
